assets/js/admin/admin.js

```diff
@@ -0,0 +1,117 @@
+     * @param data - The SVG contents.
+     */
+    const ajaxUrlParams = (svgUrl, data) => {
+        // Run the SVGO optimizer to get the optimized SVG contents.
+        const optimized = optimize(data, svgoParams);
+        const optimizedString = optimized?.data;
+        // Do not proceed if the optimized string is empty or the same as the initial data, and therefore already optimized.
+        if (!optimizedString || (data === optimizedString)) {
+            return null;
+        }
+
+        // Prepare the parameters for the AJAX Call.
+        return {
+            action: 'safe_svg_optimize',
+            svg_url: svgUrl,
+            optimized_svg: optimizedString,
+            svg_nonce: safeSvgParams.nonce
+        };
+    };
+
+    /**
+     * Hook into the WordPress Uploader and optimize the SVG.
+     */
+    Object.assign(
+        wp.Uploader.prototype, {
+            // Run on a successful upload.
+            success: function (attachment) {
+
+                const svgUrl = attachment?.attributes?.url;
+                if (!svgUrl || 'svg+xml' !== attachment?.attributes?.subtype) {
+                    return;
+                }
+
+                // Get the SVG data from the file's URL and optimize.
+                fetch(svgUrl, {method: 'GET'})
+                    .then((response) => response.text())
+                    .then((response) => {
+
+                        const params = ajaxUrlParams(svgUrl, response);
+                        if (!params) {
+                            return;
+                        }
+                        // Make an AJAX call to update the SVG file with the optimized contents.
+                        ajaxUrl.search = new URLSearchParams(params);
+                        fetch(ajaxUrl, {method: 'GET'})
+                            .then((ajaxResponse) => ajaxResponse)
+                            .then((ajaxResponse) => {
+                                if (200 !== ajaxResponse?.status) {
+                                    return;
+                                }
+
+                                // Refresh the uploader window to update the file size.
+                                if (wp.media.frame.content.get() !== null) {
+                                    wp.media.frame.content.get().collection.props.set({ignore: (+new Date())});
+                                    wp.media.frame.content.get().options.selection.reset();
+                                } else {
+                                    wp.media.frame.library.props.set({ignore: (+new Date())});
+                                }
+
+                            });
+                    });
+            }
+        }
+    );
+
+    /**
+     * Optimize the SVGs inserted in the editor.
+     * This takes care of SVGs uploaded via direct upload, without using the Media Library.
+     */
+    const {isSavingPost, getPostEdits} = select('core/editor');
+
+    subscribe(() => {
+        if (isSavingPost()) {
```

Comment:
**thought (non-blocking):** The handling of this on save concerns me a little. We're making 2 additional requests for each image on the page. That could add up quickly and really bog down the save process of the block editor. 

Have we looked other options, such as something like https://github.com/spatie/image-optimizer which _may_ allow us to do the optimisation from PHP at upload time.

Reply:
@darylldoyle I agree with your arguments, and this is something that concerned me as well before I ended up with the specific approach. When testing the `image-optimizer`, though, I came across the following issues, which made me avoid it:
1. It uses SGVO 1, which is the older version of SVGO, and based on that [comment](https://github.com/spatie/image-optimizer/issues/147), it seems to have a vulnerable dependency.
2. When I tried to test it, I got a fatal error, as one of the tool's dependencies seem to require PHP 8.1.0. (The error was `Your Composer dependencies require a PHP version ">= 8.1.0".`. 

I do agree, that a PHP-based approach would be ideal, and much simpler to implement. I didn't find a reliable tool to use instead, though.